python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -260,9 +448,61 @@ def __init__(
             instance=instance,
         )
 
+    def _cache_initial_unconsumed_events(self) -> None:
+        from dagster._core.events import DagsterEventType
+        from dagster._core.storage.event_log.base import EventRecordsFilter
+
+        # This method caches the initial unconsumed events for each asset key. To generate the
+        # current unconsumed events, call get_unconsumed_events instead.
+        if self._fetched_initial_unconsumed_events:
+            return
+
+        for asset_key in self._asset_keys:
+            event_records = self.instance.get_event_records(
+                EventRecordsFilter(
+                    event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                    storage_ids=list(
+                        self._get_cursor(
+                            asset_key
+                        ).trailing_unconsumed_partitioned_event_ids.values()
+                    ),
+                )
+            )
+            self._initial_unconsumed_events_by_id.update(
+                {event_record.storage_id: event_record for event_record in event_records}
+            )
+
+        self._fetched_initial_unconsumed_events = True
+
+    def _get_unconsumed_events_with_ids(self, event_ids: List[int]) -> Sequence["EventLogRecord"]:
+        self._cache_initial_unconsumed_events()
+        unconsumed_events = []
+        for event_id in event_ids:
+            event = self._initial_unconsumed_events_by_id.get(event_id)
+            unconsumed_events.extend([event] if event else [])
+
+        return unconsumed_events
+
+    @public
```

Comment:
Is there a situation where we imagine that a user would need to use this? If not, could be best to hold it back? If so, I think we should make sure that the name makes it clear these are _trailing_ unconsumed events (if I'm understanding correctly that they're only the trailing ones).

Reply:
Hmm, so I included this method so that users would have a way to mark the unconsumed events as consumed when they reach the max num allowed unconsumed events. Otherwise, there aren't any built in ways for them to fetch the unconsumed events. 

I can update this name to be trailing unconsumed events instead!